src/volatile_memory.rs

```diff
@@ -671,6 +679,223 @@ impl<'a, T: ByteValued> VolatileRef<'a, T> {
+///
+/// ```
+/// # use vm_memory::VolatileRef;
+///   let mut v = 5u32;
+///   assert_eq!(v, 5);
+///   let v_ref = unsafe { VolatileRef::<u32>::new(&mut v as *mut u32 as *mut u8) };
+///   assert_eq!(v_ref.load(), 5);
+///   v_ref.store(500);
+///   assert_eq!(v, 500);
+#[derive(Debug)]
+pub struct VolatileArrayRef<'a, T: ByteValued>
+where
+    T: 'a,
+{
+    addr: *mut u8,
+    nelem: usize,
+    phantom: PhantomData<&'a T>,
+}
+
+impl<'a, T: ByteValued> VolatileArrayRef<'a, T> {
+    /// Creates a reference to raw memory that must support volatile access of `T` sized chunks.
+    ///
+    /// To use this safely, the caller must guarantee that the memory at `addr` is big enough for
+    /// `nelem` values of type `T` and is available for the duration of the lifetime of the new
+    /// `VolatileRef`. The caller must also guarantee that all other users of the given chunk of
+    /// memory are using volatile accesses.
+    pub unsafe fn new(addr: *mut u8, nelem: usize) -> VolatileArrayRef<'a, T> {
+        VolatileArrayRef {
+            addr,
+            nelem,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Returns whether this slice is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use vm_memory::{VolatileRef, VolatileArrayRef};
+    ///   let v_array = unsafe { VolatileArrayRef::<u32>::new(0 as *mut _, 0) };
+    ///   assert!(v_array.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.nelem == 0
+    }
+
+    /// Gets the size of this slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use vm_memory::{VolatileRef, VolatileArrayRef};
+    ///   let v_array = unsafe { VolatileArrayRef::<u32>::new(0 as *mut _, 1) };
+    ///   assert_eq!(v_array.len(), 1);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.nelem
+    }
+
+    /// Gets the element size of this slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::mem::size_of;
+    /// # use vm_memory::VolatileRef;
+    ///   let v_ref = unsafe { VolatileRef::<u32>::new(0 as *mut _) };
+    ///   assert_eq!(v_ref.len(), size_of::<u32>() as usize);
+    /// ```
+    pub fn element_size(&self) -> usize {
+        size_of::<T>()
+    }
+
+    /// Gets the address of this slice's memory.
+    pub fn as_ptr(&self) -> *mut u8 {
+        self.addr
+    }
+
+    /// Converts this `T` reference to a raw slice with the same size and address.
+    pub fn to_slice(&self) -> VolatileSlice<'a> {
+        unsafe { VolatileSlice::new(self.addr, self.nelem * self.element_size()) }
+    }
+
+    /// Does a volatile read of the I-th value pointed to by this ref.
+    pub fn ref_at(&self, index: usize) -> VolatileRef<'a, T> {
+        // Safe because the memory has the same lifetime and points to a subset of the
+        // memory of the VolatileArrayRef.
+        unsafe {
+            let byteofs = (self.element_size() * index) as isize;
+            let ptr = self.as_ptr().offset(byteofs);
```

Comment:
I still don't understand how `ptr::offset` and `ptr::wrapping_offset` work/differ from each other :(, but the former does state *the computed offset, in bytes, cannot overflow an isize*, otherwise undefined behavior might happen. Is that a concern here?

Reply:
I'll add a check against `isize` in `get_array_ref`. And especially, we need to range-check here!